libraries/Microsoft.Bot.Builder.Expressions/CommonRegex.cs

```diff
@@ -1,24 +1,59 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.Text;
+using System.Linq;
 using System.Text.RegularExpressions;
 using Antlr4.Runtime;
 using Antlr4.Runtime.Misc;
 using Antlr4.Runtime.Tree;
 
 namespace Microsoft.Bot.Builder.Expressions
 {
+    internal class RegexLRUEntity
+    {
+        public RegexLRUEntity(long timestamp, Regex regex)
+        {
+            this.Timestamp = timestamp;
+            this.Regex = regex;
+        }
+
+        /// <summary>
+        /// Gets time when this item inserted.
+        /// </summary>
+        /// <value>
+        /// Time when this item inserted.
+        /// </value>
+        public long Timestamp { get; }
+
+        /// <summary>
+        /// Gets regex entity of cache.
+        /// </summary>
+        /// <value>
+        /// Regex entity of cache.
+        /// </value>
+        public Regex Regex { get; }
+    }
+
     public class CommonRegex
     {
+        private const int CacheCapacity = 15;
+        private static Dictionary<string, RegexLRUEntity> lruCacheMap = new Dictionary<string, RegexLRUEntity>();
+        private static readonly object LockObj = new object();
+
         public static Regex CreateRegex(string pattern)
         {
             if (string.IsNullOrEmpty(pattern) || !IsCommonRegex(pattern))
             {
                 throw new ArgumentException("A regular expression parsing error occurred.");
             }
 
-            // TODO check pattern
-            return new Regex(pattern);
+            var regex = GetCacheValue(pattern);
```

Comment:
Cache should be called before IsCommonRegex, because you already parse regex in IsCommonRegex, that make cache less useful